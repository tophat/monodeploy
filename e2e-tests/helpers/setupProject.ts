/* eslint-disable jest/no-export */
import { promises as fs } from 'fs'
import os from 'os'
import path from 'path'

import { type ExecException, exec } from '@monodeploy/io'
import {
    addGitRemote,
    cleanUp,
    initGitRepository,
    setupTestRepository,
    writeConfig,
} from '@monodeploy/test-utils'
import { type MonodeployConfiguration, type RecursivePartial } from '@monodeploy/types'
import { npath } from '@yarnpkg/fslib'

import { startRegistry, stopRegistry, waitForRegistry } from './docker'
import run from './runner'

const registryUrl = 'http://localhost:4873'

type RunFn = (args?: string[]) => Promise<{
    stdout: string | undefined
    stderr: string | undefined
    error?: Error | ExecException
}>

type ExecFn = (cmd: string) => ReturnType<typeof exec>

type ReadFile = (filepath: string) => Promise<string>
type WriteFile = (filepath: string, data: string | Record<string, unknown>) => Promise<string>

type TestCase = (params: {
    cwd: string
    run: RunFn
    readFile: ReadFile
    writeFile: WriteFile
    exec: ExecFn
}) => Promise<void>

export default function setupProject({
    repository,
    config,
    testCase,
}: {
    repository: Parameters<typeof setupTestRepository>
    config: RecursivePartial<MonodeployConfiguration>
    testCase: TestCase
}): () => Promise<void> {
    return async (): Promise<void> => {
        try {
            await stopRegistry()
        } catch {}
        await startRegistry()
        await waitForRegistry(25000)

        let project: string | null = null
        let remotePath: string | null = null

        try {
            // the project we're publishing
            project = await setupTestRepository(...repository)

            // remote to push tags/artifacts to
            remotePath = await fs.mkdtemp(path.join(os.tmpdir(), 'monorepo-'))
            await initGitRepository(npath.toPortablePath(remotePath))
            await addGitRemote(project, remotePath, 'origin')

            const configFilename = path.relative(
                project,
                await writeConfig({
                    cwd: project,
                    config: {
                        registryUrl,
                        ...config,
                    },
                }),
            )

            // initial commit
            await exec('git pull --rebase --no-verify origin main', {
                cwd: project,
            })
            await exec('git add . && git commit -n -m "initial commit"', {
                cwd: project,
            })
            await exec('git push -u origin main', {
                cwd: project,
            })

            await testCase({
                cwd: project,
                run: (args?: string[]) => {
                    if (!project) throw new Error('Missing project path.')
                    console.log(`Temporary Project: ${project}`)

                    return run({
                        cwd: project,
                        args: [`--config-file ${configFilename}`, ...(args ? args : [])].join(' '),
                    })
                },
                readFile: (filename: string) => {
                    if (!project) throw new Error('Missing project path.')
                    return fs.readFile(path.resolve(project, filename), {
                        encoding: 'utf8',
                    })
                },
                writeFile: async (
                    filename: string,
                    data: string | Record<string, unknown>,
                ): Promise<string> => {
                    if (!project) throw new Error('Missing project path.')
                    const fullFilename = path.resolve(project, filename)
                    await fs.appendFile(
                        fullFilename,
                        typeof data === 'string' ? data : JSON.stringify(data, null, 4),
                        'utf-8',
                    )
                    return fullFilename
                },
                exec: (command: string) => {
                    if (!project) throw new Error('Missing project path.')
                    return exec(command, { cwd: project })
                },
            })
        } finally {
            await cleanUp([project, remotePath].filter((v): v is string => v !== null))
            await stopRegistry()
        }
    }
}
