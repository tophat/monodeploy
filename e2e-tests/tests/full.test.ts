import childProcess from 'child_process'
import util from 'util'

const exec = util.promisify(childProcess.exec)

import setupProject from 'helpers/setupProject'

const TIMEOUT = 120000 // we need time for docker interactions

describe('Full E2E', () => {
    it(
        'runs the full monodeploy pipeline',
        setupProject({
            repository: [
                {
                    'pkg-1': {},
                    'pkg-2': {
                        dependencies: ['pkg-1'],
                    },
                    'pkg-3': { dependencies: ['pkg-2'] },
                    'pkg-4': { dependencies: ['pkg-3'] },
                    'pkg-isolated': {},
                },
            ],
            config: {
                access: 'public',
                changelogFilename: 'changelog.md',
                changesetFilename: 'changes.json',
                dryRun: true,
                autoCommit: true,
                autoCommitMessage: 'chore: release',
                git: {
                    push: true,
                    remote: 'origin',
                    tag: true,
                },
                jobs: 1,
                persistVersions: true,
                noRegistry: false,
                topological: true,
                topologicalDev: true,
                maxConcurrentReads: 1,
                maxConcurrentWrites: 1,
            },
            testCase: async ({ cwd, run, readFile }) => {
                // First semantic commit
                await exec(`echo "Modification." >> packages/pkg-1/README.md`, {
                    cwd,
                })
                await exec(
                    `git add . && git commit -n -m "feat: some fancy addition" && git push`,
                    { cwd },
                )

                const { stdout, stderr, error } = await run()

                expect(error).toBeUndefined()
                console.log(stdout)
                // Locally
                const localChangeset = JSON.parse(
                    await readFile('changes.json'),
                )
                expect(localChangeset).toMatchSnapshot()
                //  TODO: snapshot changeset

                // On Remote:
                //  TODO: assert git tags
                //  TODO: assert changelog updated

                // -----

                // Make another semantic change
                // TODO: run again

                // Locally
                //  TODO: snapshot changeset

                // On Remote:
                //  TODO: assert git tags
                //  TODO: assert changelog updated
            },
        }),
        TIMEOUT,
    )
})
