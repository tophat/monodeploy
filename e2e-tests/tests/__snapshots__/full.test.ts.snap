// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Full E2E runs the full monodeploy pipeline 1`] = `
Object {
  "pkg-1": Object {
    "changelog": null,
    "tag": "pkg-1@0.1.0",
    "version": "0.1.0",
  },
  "pkg-2": Object {
    "changelog": null,
    "tag": "pkg-2@0.0.1",
    "version": "0.0.1",
  },
  "pkg-3": Object {
    "changelog": null,
    "tag": "pkg-3@0.0.1",
    "version": "0.0.1",
  },
}
`;
